src/main/java/org/folio/event/handler/HoldingUpdateAsyncRecordHandler.java

```diff
@@ -0,0 +1,193 @@
+    if (CollectionUtils.isEmpty(poLines)) {
+      log.warn("updatePoLines:: No POLs were found for holding to update, holdingId: {}", holdingId);
+      return Future.succeededFuture(List.of());
+    }
+    var updateInstanceId = updatePoLinesInstanceId(resourceEvent, poLines);
+    var searchLocationIds = updatePoLinesSearchLocationIds(resourceEvent, poLines);
+    if (!updateInstanceId && !searchLocationIds) {
+      log.warn("updatePoLines:: No POLs were updated for holding, holdingId: {}, POLs retrieved: {}", holdingId, poLines.size());
+      return Future.succeededFuture(List.of());
+    }
+    return poLinesService.updatePoLines(poLines, conn, tenantId)
+      .map(v -> {
+        log.info("updatePoLines:: Successfully updated POLs for holdingId: {}, POLs updated: {}", holdingId, poLines.size());
+        // Very important to return a null poLine array in case no instanceId update
+        // took place to avoid a recursive invocation of the same consumer
+        return updateInstanceId ? poLines : List.of();
+      });
+  }
+
+  private List<String> extractDistinctAdjacentHoldingsToUpdate(String excludingHoldingId, List<PoLine> poLines) {
+    return poLines.stream()
+      .map(PoLine::getLocations)
+      .flatMap(Collection::stream)
+      .map(Location::getHoldingId)
+      .filter(Objects::nonNull)
+      .filter(holdingId -> !holdingId.equals(excludingHoldingId))
+      .distinct()
+      .toList();
+  }
+
+  private Future<Void> updateAdjacentHoldingsWithNewInstanceId(ResourceEvent resourceEvent, Map<String, String> headers,
+                                                               List<String> holdingIds) {
+    if (CollectionUtils.isEmpty(holdingIds)) {
+      log.warn("updateAdjacentHoldingsWithNewInstanceId:: No adjacent holdings were found to update, ignoring update");
+      return Future.succeededFuture();
+    }
+    var poLineFutures = new ArrayList<Future<Void>>();
+    var requestHeaders = new HeadersMultiMap().addAll(headers);
+    var okapiUrl = headers.get(XOkapiHeaders.URL);
+    var newInstanceId = JsonObject.mapFrom(resourceEvent.getNewValue()).getString(INSTANCE_ID);
+    holdingIds.forEach(holdingId -> poLineFutures.add(updateAdjacentHolding(okapiUrl, requestHeaders, holdingId, newInstanceId)));
+    return GenericCompositeFuture.all(poLineFutures)
+      .onComplete(asyncResult -> log.info("updateAdjacentHoldingsWithNewInstanceId:: Updated adjacent holdings, size: {}", holdingIds.size()))
+      .mapEmpty();
+  }
+
+  private Future<Void> updateAdjacentHolding(String okapiUrl, MultiMap requestHeaders, String holdingId, String newInstanceId) {
+    log.info("updateAdjacentHolding:: Updating an adjacent holding, holdingId: {}, instanceId: {}", holdingId, newInstanceId);
+    var absoluteUri = String.format(STORAGE_HOLDING_URL, okapiUrl, holdingId);
+    return webClient.requestAbs(HttpMethod.GET, absoluteUri)
```

Comment:
Use RestClient for doing such calls to keep abstaction like Service -> RestClient -> WebClient

Reply:
looks good